keras_cv/layers/feature_pyramid.py

```diff
@@ -0,0 +1,208 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+
+# TODO(scottzhu): Register it later due to the conflict in the retina_net
+# @tf.keras.utils.register_keras_serializable(package="keras_cv")
+class FeaturePyramid(tf.keras.layers.Layer):
+    """Implements a Feature Pyramid Network.
+
+    This implements the paper:
+      Tsung-Yi Lin, Piotr Dollar, Ross Girshick, Kaiming He, Bharath Hariharan, and
+      Serge Belongie.
+      Feature Pyramid Networks for Object Detection.
+      (https://arxiv.org/pdf/1612.03144)
+
+    Feature Pyramid Networks (FPNs) are basic components that are added to an
+    existing CNN to extract features at different scales. For the basic FPN, the
+    inputs are features `Ci` from different levels of a CNN, where the feature is
+    scaled from the image size by `1/2^i` for any level `i`.
+
+    There is an output associated with each level in the basic FPN. The output Pi
+    at level `i` (corresponding to Ci) is given by performing a merge operation on
+    the outputs of:
+
+    1) a lateral operation on Ci (usually a conv2D layer with kernel = 1 and strides = 1)
+    2) a top-down upsampling operation from Pi+1 (except for the top most level)
+
+    The final output of each level  will also have a conv2D operation
+    (usually with kernel = 3 and strides = 1).
+
+    The inputs to the layer should be a dict with int keys should match the
+    pyramid_levels, e.g. for `pyramid_levels` = [2,3,4,5], the expected input dict should
+    be `{2:c2, 3:c3, 4:c4, 5:c5}`.
+
+    The output of the layer will have same structures as the inputs, a dict with int keys
+    and value for each of the level.
+
+    Args:
+        pyramid_levels: a sorted python int list that specifies all the values of level
+            `i` for which the feature Ci will be specified.
+        num_channels: an integer representing the number of channels for the FPN
+            operations. Defaults to 256.
+        lateral_layers: a python dict with int keys that matches to each of the pyramid
+            level. The values of the dict should be `keras.Layer`, which will be called
+            with feature inputs from backbone at each level. Default to None, and a
+            `keras.Conv2D` layer with kernel 1x1 will be created for each pyramid level.
+        output_layers: a python dict with int keys that matches to each of the pyramid
```

Comment:
when does it need to be op and when does it need to be layer? (Asking this as users might get confused)